js/src/components/paid-ads/convertToAssetGroupUpdateBody.test.js

```diff
@@ -0,0 +1,450 @@
+				{
+					id: genId(),
+					content: 'https://example/logo-1.png',
+				},
+				{
+					id: genId(),
+					content: 'https://example/logo-2.png',
+				},
+				{
+					id: genId(),
+					content: 'https://example/logo-3.png',
+				},
+			],
+			[ ASSET_KEY.HEADLINE ]: [
+				{
+					id: genId(),
+					content: 'headline 1',
+				},
+				{
+					id: genId(),
+					content: 'headline 2',
+				},
+				{
+					id: genId(),
+					content: 'headline 3',
+				},
+			],
+			[ ASSET_KEY.LONG_HEADLINE ]: [
+				{
+					id: genId(),
+					content: 'long headline 1',
+				},
+			],
+			[ ASSET_KEY.DESCRIPTION ]: [
+				{
+					id: genId(),
+					content: 'description 1',
+				},
+			],
+			[ ASSET_KEY.CALL_TO_ACTION_SELECTION ]: {
+				id: genId(),
+				content: 'sign_up',
+			},
+		},
+	};
+
+	const filledValues = {
+		[ ASSET_FORM_KEY.BUSINESS_NAME ]: 'General Clothing',
+		[ ASSET_FORM_KEY.MARKETING_IMAGE ]: [ 'https://example/image-1.png' ],
+		[ ASSET_FORM_KEY.SQUARE_MARKETING_IMAGE ]: [
+			'https://example/square-image-1.png',
+		],
+		[ ASSET_FORM_KEY.PORTRAIT_MARKETING_IMAGE ]: [
+			'https://example/portrait-image-1.png',
+		],
+		[ ASSET_FORM_KEY.LOGO ]: [
+			'https://example/logo-1.png',
+			'https://example/logo-2.png',
+			'https://example/logo-3.png',
+		],
+		[ ASSET_FORM_KEY.HEADLINE ]: [
+			'headline 1',
+			'headline 2',
+			'headline 3',
+		],
+		[ ASSET_FORM_KEY.LONG_HEADLINE ]: [ 'long headline 1' ],
+		[ ASSET_FORM_KEY.DESCRIPTION ]: [ 'description 1' ],
+		[ ASSET_FORM_KEY.CALL_TO_ACTION_SELECTION ]: 'sign_up',
+	};
+
+	function applyOperations( baseAssetGroup, operations ) {
+		const nextAssets = structuredClone( baseAssetGroup.assets );
+
+		operations.forEach( ( operation ) => {
+			const { field_type: key, id, content } = operation;
+			const isMultiple = Array.isArray( filledAssetGroup.assets[ key ] );
+			let asset = nextAssets[ key ];
+
+			if ( id === null ) {
+				if ( isMultiple ) {
+					if ( ! asset ) {
+						asset = [];
+					}
+					asset.push( { content } );
+				} else {
+					asset = { content };
+				}
+			} else if ( content === null ) {
+				if ( isMultiple ) {
+					asset = asset.filter( ( item ) => item.id !== id );
+				} else {
+					asset = { content: '' };
+				}
+			}
+
+			nextAssets[ key ] = asset;
+		} );
+
+		const appliedValues = {};
+
+		Object.entries( nextAssets ).forEach( ( [ key, asset ] ) => {
+			if ( Array.isArray( asset ) ) {
+				appliedValues[ key ] = asset.map( ( { content } ) => content );
+			} else {
+				appliedValues[ key ] = asset.content;
+			}
+		} );
+
+		return appliedValues;
+	}
+
+	let assetGroup;
+	let values;
+
+	beforeEach( () => {
+		assetGroup = structuredClone( filledAssetGroup );
+		values = structuredClone( filledValues );
+	} );
+
+	test( 'in each operation, one of `id` and `content` must be null and the other must not be null', () => {
+		// If this test failed, the modified code might be problematic or
+		// need adjustment for corresponding tests.
```

Comment:
💅  What does this mean?  🤔

I guess this applies to all the tests.

Reply:
This test case verifies the contractual rules with the backend API, so if it fails, it will need to be confirmed carefully. I rephrased the code comment to make the reason clearer. See https://github.com/woocommerce/google-listings-and-ads/pull/1874/commits/c155aec7ec2d9a87a77d55c8709cc591ab8638c9.